@Oppen 

I don't think that they are cancelled. I just think that we must have a fast-fail and that they stop after the initial failure...

```
CRoaring 0.2.65
Alloc = 4 MiB	TotalAlloc = 41 MiB	Sys = 70 MiB	NumGC = 12
Alloc = 3 MiB	TotalAlloc = 80 MiB	Sys = 70 MiB	NumGC = 23
Alloc = 4 MiB	TotalAlloc = 120 MiB	Sys = 70 MiB	NumGC = 34
Alloc = 3 MiB	TotalAlloc = 159 MiB	Sys = 70 MiB	NumGC = 46
Alloc = 4 MiB	TotalAlloc = 198 MiB	Sys = 70 MiB	NumGC = 57
Alloc = 3 MiB	TotalAlloc = 238 MiB	Sys = 70 MiB	NumGC = 69
Alloc = 4 MiB	TotalAlloc = 277 MiB	Sys = 70 MiB	NumGC = 81
Alloc = 3 MiB	TotalAlloc = 316 MiB	Sys = 70 MiB	NumGC = 92
Alloc = 4 MiB	TotalAlloc = 356 MiB	Sys = 70 MiB	NumGC = 104
Alloc = 3 MiB	TotalAlloc = 395 MiB	Sys = 70 MiB	NumGC = 115
ok
size in kB of one bitmap  1224 ; number of copies =  7 ; total alloc:  8573 kB
TestExplicitFree output:
Starting execution... max RSS: 25636 bytes
Created reference bitmap... max RSS: 25636 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 25636 bytes
Copied then explicitly freed 100000 bitmaps... max RSS: 25636 bytes
GC'd old bitmaps... max RSS: 25636 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 25636 bytes

--- FAIL: TestExplicitFree (28.74s)
```


Please note that nothing in the C specification allows you to believe that `free` will return the memory to the operating system.

--- a/free_test/explicit_free.go
+++ b/free_test/explicit_free.go
@@ -9,35 +9,45 @@ import (
 	"github.com/RoaringBitmap/gocroaring"
 )
 
+// Exits with code 0 on success, 1 on test failure
+// or 2 on program failure (e.g. OOM or syscall error).
+
 func main() {
+	const (
+		N = 10000
+		M = 10000
+	)
+
 	var rusage syscall.Rusage
+	var maxrss_0, maxrss_free, maxrss_gc int64
 
 	fmt.Printf("Starting execution... ")
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
-	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
+	maxrss_0 = rusage.Maxrss
+	fmt.Printf("max RSS: %d bytes\n", maxrss_0)
 
 	bitmap := gocroaring.New()
-	for i := uint32(0); i < 100000; i++ {
+	for i := uint32(0); i < M; i++ {
 		bitmap.Add(i*10)
 	}
 	fmt.Printf("Created reference bitmap... ")
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
 	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
 
-	const N = 100000
 	bitmaps := make([]*gocroaring.Bitmap, N)
 	fmt.Printf("Allocated slice for %d bitmap pointers... ", N)
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
-	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
+	maxrss_free = rusage.Maxrss
+	fmt.Printf("max RSS: %d bytes\n", maxrss_free)
 
 	for i := range bitmaps {
 		bitmaps[i] = bitmap.Clone()
@@ -46,7 +56,7 @@ func main() {
 	fmt.Printf("Copied then explicitly freed %d bitmaps... ", N)
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
 	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
 
@@ -55,7 +65,7 @@ func main() {
 	fmt.Printf("GC'd old bitmaps... ")
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
 	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
 
@@ -63,7 +73,7 @@ func main() {
 	fmt.Printf("Allocated slice for %d bitmap pointers... ", N)
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
 	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
 
@@ -75,7 +85,18 @@ func main() {
 	fmt.Printf("Copied then GC'd %d bitmaps... ", N)
 	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
 		fmt.Printf("Getrusage failed with error: %s\n", err)
+		os.Exit(2)
+	}
+	maxrss_gc = rusage.Maxrss
+	fmt.Printf("max RSS: %d bytes\n", maxrss_gc)
+
+	ballpark_free := maxrss_free - maxrss_0
+	ballpark_gc := maxrss_gc - maxrss_0
+	fmt.Printf("Used ~%d bytes with explicit free\n", ballpark_free)
+	fmt.Printf("Used ~%d bytes without explicit free\n", ballpark_gc)
+
+	if ballpark_gc < 100*ballpark_free {
+		fmt.Printf("Expected a much greater difference!\n")
 		os.Exit(1)
 	}
-	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
 }

